python/cudf/cudf/core/udf/_nrt_cuda.py

```diff
@@ -0,0 +1,148 @@
+        module, incref_decref_ty, "NRT_decref"
+    )
+    # Cannot inline this for refcount pruning to work
+    fn_decref.attributes.add("noinline")
+    calldtor = ir.Function(
+        module,
+        ir.FunctionType(ir.VoidType(), [_pointer_type]),
+        name="NRT_MemInfo_call_dtor",
+    )
+
+    builder = ir.IRBuilder(fn_decref.append_basic_block())
+    [ptr] = fn_decref.args
+    is_null = builder.icmp_unsigned(
+        "==", ptr, cgutils.get_null_value(ptr.type)
+    )
+    with cgutils.if_unlikely(builder, is_null):
+        builder.ret_void()
+
+    # For memory fence usage, see https://llvm.org/docs/Atomics.html
+
+    # A release fence is used before the relevant write operation.
+    # No-op on x86.  On POWER, it lowers to lwsync.
+    #    builder.fence("release")
+
+    word_ptr = builder.bitcast(ptr, atomic_decr.args[0].type)
+
+    if config.DEBUG_NRT:
+        cgutils.printf(
+            builder, "*** NRT_Decref %zu [%p]\n", builder.load(word_ptr), ptr
+        )
+    newrefct = builder.call(
+        atomic_decr, [word_ptr]  # nvvmutils.atomic_dec_int64, etc
+    )
+
+    refct_eq_0 = builder.icmp_unsigned(
+        "==", newrefct, ir.Constant(newrefct.type, 0)
+    )
+    with cgutils.if_unlikely(builder, refct_eq_0):
+        # An acquire fence is used after the relevant read operation.
+        # No-op on x86.  On POWER, it lowers to lwsync.
+        #        builder.fence("acquire")
+        builder.call(calldtor, [ptr])
+    builder.ret_void()
+
+
+def create_nrt_module(ctx):
+    """
+    Create an IR module defining the LLVM NRT functions.
+    A (IR module, library) tuple is returned.
+    """
+    codegen = ctx.codegen()
+    library = codegen.create_library("nrt")
+
+    # Implement LLVM module with atomic ops
+    ir_mod = library.create_ir_module("nrt_module")
+
+    atomic_inc = _define_atomic_inc_dec(ir_mod, "add", ordering="monotonic")
+    atomic_dec = _define_atomic_inc_dec(ir_mod, "sub", ordering="monotonic")
+    _define_atomic_cas(ir_mod, ordering="monotonic")
+
+    _define_nrt_meminfo_data(ir_mod)
+    _define_nrt_incref(ir_mod, atomic_inc)
+    _define_nrt_decref(ir_mod, atomic_dec)
+
+    #    _define_nrt_unresolved_abort(ctx, ir_mod)
+    return ir_mod, library
+
+
+def compile_nrt_functions(ctx):
+    """
+    Compile all LLVM NRT functions and return a library containing them.
+    The library is created using the given target context.
+    """
+    ir_mod, library = create_nrt_module(ctx)
+
+    library.add_ir_module(ir_mod)
+    library.finalize()
+
+    return library
+
+
+# enable NRT for the CUDATargetContext
+CUDATargetContext.enable_nrt = True
+# compile NRT functions written in lowering into a PTX string
+library = compile_nrt_functions(descriptor.cuda_target.target_context)
+_cuda_nrt_library = compile_nrt_functions(
+    descriptor.cuda_target.target_context
+)
+numba_cuda_runtime = _cuda_nrt_library._get_ptxes()[0].encode()
+
+
+def install_registry(self, registry):
```

Comment:
Replacing this method within numba's internals might not be necessary for this feature to work, will figure it out and post my findings here.